[Bug Fix] Save actions inside dropdown don't trigger HTML5 Validation

pxpm opened the pull request:
Fixes #2339
When selecting any option from save actions dropdown the html5 validation was not running.

`EDIT: Dunno what happened with this lang files, i just changed 1 file and commited 1 file. Never changed nothing in this lang files.`

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/c9a5f56b-e46b-4a69-ac9b-a18734cf7c17) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/a40f7311-ee9c-422f-8e27-6d3702138f4e) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/c9a5f56b-e46b-4a69-ac9b-a18734cf7c17) completed: **5 new issues**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/a40f7311-ee9c-422f-8e27-6d3702138f4e) completed: **5 new issues**

tabacitu commented on src/resources/views/crud/inc/form_save_buttons.blade.php:
```diff
@@ -13,7 +13,7 @@
             <button id="btnGroupDrop1" type="button" class="btn btn-success dropdown-toggle" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false"><span class="caret"></span><span class="sr-only">&#x25BC;</span></button>
             <div class="dropdown-menu" aria-labelledby="btnGroupDrop1">
```
I think this is a good opportunity to rename ```#btnGroupDrop1``` into something better. I'll do it.

tabacitu replied:
Done.

tabacitu commented on src/resources/views/crud/inc/form_save_buttons.blade.php:
```diff
@@ -22,3 +22,26 @@
+
+@push('after_scripts')
+<script>
+
```
Usually it's best to have at least one comment for each JS code block - so that it's fast & easy to understand what that code block does without having to go through it. I'll do it.

tabacitu replied:
Done.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/ffb48667-a766-4416-8128-bfbd9c057956) was created.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/de552d5b-a017-427e-be81-f6cb61b3d2a3) was created.

tabacitu commented on src/resources/views/crud/inc/form_save_buttons.blade.php:
```diff
@@ -22,3 +22,27 @@
+                e.stopPropagation()
+                let form = $(this).closest('form');
+                if (!form[0].checkValidity || form[0].checkValidity()) {
+                    form.submit();
```
@pxpm I'm not sure this line is the problem. But there is problem 😄 My tests show that this PR breaks the functionality for the "Save and new" and "Save and edit" buttons.

No matter what Save button you click, the functionality that gets triggered is "Save and back".

![2020-01-03 10 11 53](https://user-images.githubusercontent.com/1032474/71713262-a83b4600-2e11-11ea-9977-9ab39059429f.gif)

Let me know if I haven't explained myself correctly, I'm not sure I have.

Cheers!

pxpm replied:
Totally right @tabacitu ! Forgot to update the saveAction in the input, my bad! Guess I found the code now in [form_content.blade.php](https://github.com/Laravel-Backpack/CRUD/blob/27c5d9dd2d01b8b3ada5ba29cc587c5e546f5959/src/resources/views/crud/form_content.blade.php#L59)

Should I add this code to complete the one in `form_content` or remove the bits from there and have it in this file ?

Best,
Pedro

tabacitu replied:
Makes perfect sense to move the Save button logic here, from ```form_content```. Totally agree!

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/de552d5b-a017-427e-be81-f6cb61b3d2a3) completed: **No new issues**

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/18b41355-1613-4652-b2c8-57bb660f2f71) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/18b41355-1613-4652-b2c8-57bb660f2f71) completed: **No new issues**

pxpm commented:
Hey @tabacitu 

Fixed that bug and found another. Well technically not "our" bug. It's just the way HTML5 works.

### PROBLEM

If we have 2 tabs (TAB A and TAB B) both with 2 fields, but TAB A has one that is required. If we are on TAB A and try to submit, no errors, but if we are in TAB B and try to submit with error on required field on TAB A, we will get something along this lines in console:

`An invalid form control with name='field_name' is not focusable.`

Nothing more is displayed and nothing happens. This could be very confusing for our users.

This happens because `field_name` is on TAB A and we are on TAB B, so browser can't focus on field on another TAB, but get the error and prevent submiting the form. 

### Solution
I searched alot, and tried diferent solutions. The only that I found to avoid this is switching to the tab where first error occurs and then display the error. 

This is not the perfect solution, and unfortunatelly I think there is no perfect scenario for this. If you have required fields in different tabs you still get the error on console that the fields in the other tabs can't be focused. But we assure we show the first error to the user and switch to the tab accordingly.

Let me know your thougths on this. 

Best,
Pedro

tabacitu commented on src/resources/views/crud/inc/form_save_buttons.blade.php:
```diff
@@ -22,3 +22,74 @@
+@push('after_scripts')
+<script>
+    function checkFormValidity(form) {
+        if (!form[0].checkValidity || form[0].checkValidity()) {
```
Hmm... Wouldn't this return ```true``` always? If the condition is ```if (!form[0].checkValidity || form[0].checkValidity())``` wouldn't this be true at all times? Idk.

pxpm replied:
From my understanding the first checkValidity checks if checkValidity() is supported in browser, the second actually runs the validation. Maybe I misinterpreted the docs. But my search lead me that way.

tabacitu replied:
Ok got it, thanks!

tabacitu commented:
This looks good to me! Thank you @pxpm . I just have this one question: https://github.com/Laravel-Backpack/CRUD/pull/2340/files#r370925989

shealavington commented:
Is this still being addressed?

tabacitu commented:
As far as I know, this has been fixed in Backpack 4.1 (currently in alpha, due out very soon) in a different way. So merging this fix inside 4.0 will cause conflicts when upgrading to 4.1. Am I correct @pxpm ? 

If so, I'd rather just not include this fix in 4.0, if we launch 4.1 soon and it'll have this fixed in a better way. What do you think @pxpm ?

tabacitu commented:
I thought we already fixed this, but no, we haven't. https://github.com/Laravel-Backpack/CRUD/issues/4125 popped up, so let's fix this now, during the "v5 launch polish" phase.

@pxpm can you please fix the conflicts here, re-test this and give me a green light to merge this?

pxpm commented:
It's done @tabacitu fixed conflicts, refactored and added some more comments.

I've tested, everything seems to be working ok. 

You need to add in your field definition :

```php
// ... field definition
'attributes' => [
    'required' => 'required'
]
```
So you can test the browser validation.

tabacitu commented:
@jorgetwgroup / @maurohmartinez could you please review and test this?

tabacitu commented:
@maurohmartinez you might need to merge `main` into `fix-html5-validation` to make that error go away.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/3b7f48fb-5cc1-4bd8-b850-6e10b355ef5f) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/Laravel-Backpack/CRUD/inspections/3b7f48fb-5cc1-4bd8-b850-6e10b355ef5f) completed: **No new issues**

maurohmartinez commented:
@pxpm Tested on Chrome, Safari, and Firefox (MAC) and it worked as a charm!
To me, it looks ready @tabacitu.

jorgetwgroup commented:
Hi @pxpm is working great!

I add email field to 3rd tab on Monsters

![screenshot-backpack test-2022 11 21-17_39_03](https://user-images.githubusercontent.com/44643608/203154421-7ec78673-61a8-4575-8a39-eaa68991881b.png)

When im in the same tab, validation works

![Screenshot 2022-11-21 at 17 39 33](https://user-images.githubusercontent.com/44643608/203154506-826a2288-a130-4b74-847c-5c7c9f55a212.png)

If i go to other tab, validation work, but didnt show

![Screenshot 2022-11-21 at 17 39 55](https://user-images.githubusercontent.com/44643608/203154568-8125eec6-e053-4e20-aff1-ab066fc85b62.png)

Just receive console notification
![Screenshot 2022-11-21 at 17 40 05](https://user-images.githubusercontent.com/44643608/203154589-f96fa949-3838-4461-a998-ba0b89eed047.png)

On "fix-html5-validation" branch, try same steps and everything works great

![Screenshot 2022-11-21 at 17 40 59](https://user-images.githubusercontent.com/44643608/203154719-33c0352c-8bf6-4e06-9dae-952a2128bac5.png)

Even try to force not default save, still working.

![Screenshot 2022-11-21 at 17 41 31](https://user-images.githubusercontent.com/44643608/203154784-4f8bc4a7-6f25-4465-9479-266662562dbb.png)

Is ready to merge.

Cheers.

jorgetwgroup commented:
Im sorry @maurohmartinez didnt see your answer before 😳 i dont want to make work twice.

Will not happen again.
